Documentation - Basic Construction (#86)
cardano-scaling/alba: docs/basic_construction.md

# Telescope - Basic Construction

> **ALBA** provides a method for a prover to convince a verifier that the prover knows at more than $n_f$ elements satisfying a predicate $R$, while only presenting a small subset of elements as proof. 

Reviewer: ```suggestion
**ALBA** provides a method for a prover to convince a verifier that the prover knows stricly more than $n_f$ elements satisfying a predicate $R$, while only presenting a smaller subset of elements as proof. 
```

This construction uses a recursive filtering process, guided by random oracles, to produce succinct and efficient proofs.

> ## Overview

Reviewer: Perhaps the first and last sentences could be outside the itemize and the rest in a numbered list?
It feels like you are cramming together different things: how the protocol works (step by step) and information about the oracle and so on.

- The prover builds a proof by creating tuples (sequences) of elements from $S_p$, the set of elements. 

> - The process begins with roots as starting points and expands them step-by-step. 

Reviewer: Perhaps rewrite this with "The tuple are initilised with ..." ?
The term "roots" here is a bit confusing as you haven't talked about trees yet.

> - At each step, only tuples that satisfy specific conditions, as determined by a random oracle, are kept for further extension.

Reviewer: We only check the tuple once it is updated. Perhaps you could write something like:
"At each step, the prover takes a tuple, extends it with an element of Sp, and perform an oracle check. If the check is successful, the tuple is kept, and may be extended further, otherwise it is discarded."

> - The random oracle acts as a filter to ensure that the prover selects only a small subset of elements that is highly unlikely to satisfy the conditions by chance alone.

Reviewer: I am not sure about what you want to mean here?

Reviewer: @curiecrypt I am not sure you resolved this.

What are the conditions if not the oracle check? Why would the RO be a filter then?

  - $u: ~~$ The length of the proof tuple generated by the prover.
  - $q: ~~$ The probability that a tuple of full size is selected.
  - $n_p: ~~$ Total number of elements available to the prover.
  - $n_f: ~~$ The lower bound to prove on prover set.

- **Random Oracles**:
  - $H_1: ~~$ A random oracle returning $1$ with probability $1/n_p$, used to probabilistically filter candidate tuples. 
  - $H_2: ~~$ Another random oracle returning $1$ with probability $q$, applied as a final test to determine if a tuple qualifies as a valid proof.

### The Tree Structure
- Each root $t \in \[d\]$ is treated as the starting node of a tree.
- The tree has $u$ levels, where $u$ is the length of the proof tuple.
- At level $i$, nodes represent tuples of the form $(t, s_1, s_2, \ldots, s_i)$, where $t$ is the root, and $s_1, \ldots, s_i$ are elements.

### Tuple Construction via DFS
- The prover constructs tuples recursively using DFS.
- At each level $i$, the prover:
  - Extends the current tuple by adding an element $s_{i+1}$ from $S_p$.
  - Checks tuple validity using the random oracle $H_1$. A tuple $(t, s_1, \ldots, s_{i+1})$ is valid if $H_1(t, s_1, \ldots, s_{i+1}) = 1$.

>   - If the tuple fails the $H_1$ check, it is discarded, and the DFS does not explore its descendants.

Reviewer: Add that otherwise we call dfs on the newly extended tuple

- The process continues until a valid tuple of length $u$ is found, or all possibilities are exhausted.

#### Example: how DFS works
- Choose a root $t \in \[d\]$ and initialize the search with $(t)$.

>   - At each step, extend the current tuple $(t, s_1, \ldots, s_i)$ by trying every element $s_{i+1} \in S_p$.

Reviewer: Why not iterate on s \in Sp directly? "for s in Sp: ..."

Reviewer: @curiecrypt I am not sure you resolved this (perhaps you discarded this suggestion?)

Author: I changed other occurences but I totally missed this one. I am on it.

  - Check if $H_1(t, s_1, \ldots, s_{i+1}) = 1$:

>     - If true, recursively extend the tuple.

Reviewer: call dfs on new tuple you mean?

    - If false, backtrack and try the next candidate for $s_{i+1}$.
- Stop when a valid tuple $(t, s_1, \ldots, s_u)$ is found and satisfies $H_2(t, s_1, \ldots, s_u) = 1$.

> ## The Protocol

Reviewer: Should you add a parameter generation algorithm? I think it would be good to write that $u$ is a function of $l_sec / log(n_p/n_f)$

### Prover Algorithm

> The prover's objective is to convince the verifier that they know a large subset of elements $S_p$ without revealing the entire set.

Reviewer: I would add the threshold, and say without revealing neither the entire subset or any subset of size higher than nf.

Instead, the prover constructs a compact proof $(t, s_1, ..., s_u)$ that satisfies the following:
- The proof demonstrates that the prover possesses knowledge of a set $S_p$ satisfying a predicate $R$, with the assurance that $|S_p| > n_f$, where $n_f$ is a threshold strictly smaller than $n_p$ and $|S_p| \geq n_p$.
- The prover outputs only a small sequence of $u$ elements from $S_p$, rather than transmitting the full set, ensuring communication efficiency.
- By using random oracles $H_1$ and $H_2$, the prover ensures that:
  - The proof is hard to forge by an adversary who knows at most $n_f$ elements.

>   - Only the prover, with access to a large $S_p$, can construct a valid proof with high probability.

Reviewer: I am not sure what you are trying to convey here. The prover needs to know a "large" Sp?

Reviewer: It seems you are trying to define completeness but you are adding a condition on Sp on it

- The constructed tuple $(t, s_1, ..., s_u)$ satisfies:
  - $H_1(t, s_1, ..., s_i) = 1$ for all prefixes up to length $u$,
  - $H_2(t, s_1, ..., s_u) = 1$ for the final tuple.

In summary, the prover's task is to use recursive construction via DFS to find and output a valid tuple that demonstrates knowledge of $S_p$ without directly revealing the full set.

### Verifier Algorithm
The verifier's objective is to validate the prover's claim by ensuring that the submitted proof $(t, s_1, ..., s_u)$ satisfies the conditions of the ALBA protocol. Specifically:

> - Verify that the proof tuple satisfies the required conditions of the random oracles $H_1$ and $H_2$.

Reviewer: ```suggestion
- Verify that the proof tuple satisfies the random oracles' checks for both $H_1$ and $H_2$.
```

>   - Ensure that the proof adheres to the structure specified in the protocol.

Reviewer: ```suggestion
 - Check that the proof is well formed, as specified in the protocol.
```

  - Check if $H_1(t, s_1, \ldots, s_{i+1}) = 1$:
    - If true, recursively extend the tuple.
    - If false, backtrack and try the next candidate for $s_{i+1}$.
- Stop when a valid tuple $(t, s_1, \ldots, s_u)$ is found and satisfies $H_2(t, s_1, \ldots, s_u) = 1$.

## The Protocol

### Prover Algorithm
The prover's objective is to convince the verifier that they know a large subset of elements $S_p$, with size exceeding the threshold $n_f$, without revealing the entire set or any specific subset of size greater than $n_f$.
Instead, the prover constructs a compact proof $(t, s_1, ..., s_u)$ that satisfies the following:
- The proof demonstrates that the prover possesses knowledge of a set $S_p$ satisfying a predicate $R$, with the assurance that $|S_p| > n_f$, where $n_f$ is a threshold strictly smaller than $n_p$ and $|S_p| \geq n_p$.
- The prover outputs only a small sequence of $u$ elements from $S_p$, rather than transmitting the full set, ensuring communication efficiency.
- By using random oracles $H_1$ and $H_2$, the prover ensures that:
  - The proof is hard to forge by an adversary who knows at most $n_f$ elements from $S_p$.

>   - A prover who knows a subset $S_p$ with size greater than $n_f$ can construct a valid proof with high probability, leveraging the filtering and validation mechanisms of $H_1$ and $H_2$.

Reviewer: ```suggestion
  - A prover who knows a subset $S_p$ with size greater than $n_p$ can construct a valid proof with high probability, leveraging the filtering and validation mechanisms of $H_1$ and $H_2$.
```

  - If the tuple fails the $H_1$ check, it is discarded, and the DFS does not explore its descendants.
  - Otherwise, the prover recursively calls DFS on the newly extended tuple to continue the construction process.
- The process continues until a valid tuple of length $u$ is found, or all possibilities are exhausted.

#### Example: how DFS works
- Choose a root $t \in \[d\]$ and initialize the search with $(t)$.
  - At each step, extend the current tuple $(t, s_1, \ldots, s_i)$ by trying every element $s \in S_p$.
  - Check if $H_1(t, s_1, \ldots, s_i) = 1$:
    - If true, recursively extend the tuple.
    - If false, backtrack and try the next candidate for $s_i$.
- Stop when a valid tuple $(t, s_1, \ldots, s_u)$ is found and satisfies $H_2(t, s_1, \ldots, s_u) = 1$.

## The Protocol
### Parameter Generation
The parameter generation process for the basic construction is designed to ensure the protocol achieves the desired levels of *security* and *efficiency*.
The internal parameters of ALBA, such as $u$, $d$, $q$, are computed based on the security parameters, $n_f$, $n_p$ .
- Security parameters:

>   - $\lambda_{sec}: $ defines the probability that an adversary, who knows at most $n_f$ elements, can construct a valid proof. It controls the protocol's resistance to adversarial attacks by making forgery negligibly likely.

Reviewer 2: This doesn't render on github for some reason. Maybe move the right dollar sign left?

  - $\lambda_{rel}: $ defines the probability that an honest prover, with knowledge of a sufficiently large set $S_p$, can successfully generate a valid proof. It ensures the practicality of the protocol for honest participants, emphasizing the importance of feasibility in proof construction.
- Relationship between the parameters:
  - The proof length $u$ is directly proportional to the security parameters and inversely proportional to the logarithm of the ratio $n_p / n_f$. 
    - This means that as the prover's set size $n_p$ grows compared to the threshold $n_f$, the required proof length decreases logarithmically, reflecting the reduced adversarial advantage.

>       $$

Reviewer 2: this doesn't render on github

Author: When the double dollar sign is in the bullet list, it is not rendered somehow!
So, I put all 3 formulas at the bottom.

        u \leftarrow \Bigg\lceil \frac{\lambda_{sec} + \log \lambda_{rel} + 1 - \log \log e}{\log (n_p / n_f)}\Bigg\rceil.
      $$
  - The maximum number of subtrees to search $d$ depends linearly on both the proof length $u$ and the security parameter $\lambda_{rel}$. 
    - It directly influences the ability to find a valid proof.
    - Increasing $d$ improves the chances of constructing a valid proof, ensuring robustness in the presence of random oracle checks.
      $$

>         d \leftarrow \Bigg\lceil \frac{2 u \lambda_{rel}}{\log e}\Bigg\rceil.

Reviewer 2: nit: I think the left arrow is used for randomize computations. Maybe use := ? `\coloneqq` might work.

Author: You're right, it's a copy-paste leftover.

    - Since $d$ is linked to $u$, $q$ is approximately inversely proportional to $u$.
      $$
        q \leftarrow \frac{2 \lambda_{rel}}{d \log e}.
      $$
    - A smaller $q$ reduces the chance of accepting invalid proofs, enhancing security, but also decreases the probability of finding a valid proof, necessitating a larger $d$.

### Prover Algorithm
The prover's objective is to convince the verifier that they know a large subset of elements $S_p$, with size exceeding the threshold $n_f$, without revealing the entire set or any specific subset of size greater than $n_f$.
Instead, the prover constructs a compact proof $(t, s_1, ..., s_u)$ that satisfies the following:
- The proof demonstrates that the prover possesses knowledge of a set $S_p$ satisfying a predicate $R$, with the assurance that $|S_p| > n_f$, where $n_f$ is a threshold strictly smaller than $n_p$ and $|S_p| \geq n_p$.
- The prover outputs only a small sequence of $u$ elements from $S_p$, rather than transmitting the full set, ensuring communication efficiency.
- By using random oracles $H_1$ and $H_2$, the prover ensures that:

>   - The proof is hard to forge by an adversary who knows at most $n_f$ elements from $S_p$.

Reviewer 2: The adversary doesn't really have S_p. This notion is only for honest provers. Suggest:
```suggestion
  - The proof is hard to forge by an adversary who knows at most $n_f$ elements.
```

  - A prover who knows a subset $S_p$ with size greater than $n_p$ can construct a valid proof with high probability, leveraging the filtering and validation mechanisms of $H_1$ and $H_2$.
- The constructed tuple $(t, s_1, ..., s_u)$ satisfies:
  - $H_1(t, s_1, ..., s_i) = 1$ for all prefixes up to length $u$,
  - $H_2(t, s_1, ..., s_u) = 1$ for the final tuple.

In summary, the prover's task is to use recursive construction via DFS to find and output a valid tuple that demonstrates knowledge of $S_p$ without directly revealing the full set.

### Verifier Algorithm
The verifier's objective is to validate the prover's claim by ensuring that the submitted proof $(t, s_1, ..., s_u)$ satisfies the conditions of the ALBA protocol. Specifically:
- Verify that the proof tuple satisfies the random oracles' checks for both $H_1$ and $H_2$.
 - Check that the proof is well formed, as specified in the protocol.

> - The verifier ensures that the proof was generated by a prover possessing a large subset $S_p$. If the prover knows fewer than $n_f$ elements, the probability of passing verification should be negligible.

Reviewer 2: ```suggestion
- The verifier ensures that the proof was generated by a prover possessing a large subset $S_p$. If the prover knows at most $n_f$ elements, the probability of passing verification should be negligible.
```

Author: I keep making the same mistake !

- The verifier performs efficient checks proportional to the proof size $u$, making the process scalable.

### Functions
#### Proving
---
> $\mathsf{Prove}(S_p) \rightarrow \pi$
> - Input:
>   - $S_p:~~$ `prover_set`, a set of elements to be proven.
> - Output:

> >   - $\pi:~~$ `proof`, a valid proof $(t, s_1, \ldots, s_u)$ or $\emptyset$.

Reviewer 2: Here and below we should probably use the "bottom" symbol instead of the empty set. `\bot`. It's usually used to denote errors.

>     - **return** $~~ \pi.$
> - **return** $~~ \emptyset.$
---

#### Depth-first search
---
> $\mathsf{DFS}(t, slist, S_p) \rightarrow \pi$
> - Input:
>   - $t:~~$ `search_counter`, index of the current subtree being searched.
>   - $slist:~~$ `element_sequence`, candidate element sequence.
>   - $S_p:~~$ `prover_set`, a set of elements to be proven.
> - Output:
>   - $\pi:~~$ `proof`, a valid proof $(t, s_1, \ldots, s_u)$ or $\emptyset$.
> ---
> - **if** $~~ \mathsf{len}(slist) == u:$
>   - **if** $~~ \mathsf{H_2}(t, slist) == 1:$
>     - $\pi \leftarrow (t, slist)$
>     - **return** $~~ \pi.$
>   - **return** $~~ \emptyset.$
> - **for** each $~~ s_{new} \in S_p:$
>   - $slist_{new} \leftarrow slist \cup \[s_{new}\]$
>   - **if** $~~ \mathsf{H_1}(t, slist_{new}) == 1:$
>     - $\pi \leftarrow \mathsf{DFS}(t, slist_{new}, S_p)$
>     - **if** $~~ \pi ~!= \emptyset:$
>       - **return** $~~ \pi.$
> - **return** $~~ \emptyset.$
---

#### Verification
---
> $\mathsf{Verify}(\pi) \rightarrow 0/1$
> 
> - Input:
>   - $\pi:~~$ `proof`, a proof of the form $(t, s_1, \ldots, s_u)$.
> - Output:
>   - $0/1$.
> ---
> - **if** $~~ t ∉  \[d\]:$
>   - **return** $~~ 0.$

> > - **for** each $~~ i = \[1,~ u\]:$

Reviewer 2: ```suggestion
> - **for** each $~~ i \in \[1,~ u\]:$
```

At each step, the random oracle performs a check, and only tuples that satisfy this check are extended further. 
This ensures that the proof construction process proceeds efficiently by discarding invalid tuples early.

> The proof construction employs *Depth-First Search (DFS)* over a conceptual tree, where each node represents a set element, and each leaf represents a potential proof. To guarantee the generation of a proof, we iterate the DFS algorithm on several trees, each marked by a counter. An Alba proof can thus be seen as the Merkle proof of a polytree where each underlying tree's root is our counter.

Reviewer 2: I think the comparison to merkle proof is very misleading.

Reviewer: Both are Merkle path from the root to a leave though? The verification method of course is different, in Merkle proof you need to have the "associated" nodes, hash them together and check the final output is the root, while in our case we do not need associated data but just perform oracle checks, still some hash but with extra comparisons, recursively.

Reviewer: Sorry I made mistake, I meant a tree path.
